### PR TITLE
Merge `feature/flat_set` into `feature/flat_map`

### PR DESCRIPTION
**Pull request by CaseyCarter**
:warning: Commit as a merge, not as a squash!

I'm going to focus on pushing `flat_map` and `flat_set` over the finish line, and we think it will be easier to eliminate redundancies and review all of the similar code if it's in one branch. This PR merges `feature/flat_set` into `feature/flat_map`, after which `feature/flat_set` will be closed. There were (unsurprisingly) a dozen or so simple merge conflicts since the two header names alphabetize together. There's also an additional commit here to eliminate the duplicate definitions of `sorted_unique` et al. by pushing them up into `<xmemory>`.

<details><summary>Output from <code>git show 0e281d02</code> (changes in the merge commit, i.e., conflict resolutions)</summary>

```
commit 0e281d02ae25fa78d6510cff439f7f25980372a6
Merge: 3bdfd744 f6805eea
Author: Casey Carter <cacarter@microsoft.com>
Date:   Thu Sep 19 17:41:40 2024 -0700

    Merge remote-tracking branch 'origin/feature/flat_set' into flat_map

diff --cc stl/CMakeLists.txt
index 6be04941,b4e0f31a..303ae13a
--- a/stl/CMakeLists.txt
+++ b/stl/CMakeLists.txt
@@@ -76,7 -76,7 +76,8 @@@ set(HEADER
      ${CMAKE_CURRENT_LIST_DIR}/inc/experimental/generator
      ${CMAKE_CURRENT_LIST_DIR}/inc/experimental/resumable
      ${CMAKE_CURRENT_LIST_DIR}/inc/filesystem
 +    ${CMAKE_CURRENT_LIST_DIR}/inc/flat_map
+     ${CMAKE_CURRENT_LIST_DIR}/inc/flat_set
      ${CMAKE_CURRENT_LIST_DIR}/inc/format
      ${CMAKE_CURRENT_LIST_DIR}/inc/forward_list
      ${CMAKE_CURRENT_LIST_DIR}/inc/fstream
diff --cc stl/inc/__msvc_all_public_headers.hpp
index 68b38d6c,d43e2d57..44170ded
--- a/stl/inc/__msvc_all_public_headers.hpp
+++ b/stl/inc/__msvc_all_public_headers.hpp
@@@ -82,7 -82,7 +82,8 @@@
  #include <exception>
  #include <expected>
  #include <filesystem>
 +#include <flat_map>
+ #include <flat_set>
  #include <format>
  #include <forward_list>
  #include <fstream>
diff --cc stl/inc/header-units.json
index 705532e8,4f74e8f9..747ea235
--- a/stl/inc/header-units.json
+++ b/stl/inc/header-units.json
@@@ -70,7 -70,7 +70,8 @@@
          "execution",
          "expected",
          "filesystem",
 +        "flat_map",
+         "flat_set",
          "format",
          "forward_list",
          "fstream",
diff --cc stl/inc/yvals_core.h
index cab9fdbd,bba7a4bf..d9076fb0
--- a/stl/inc/yvals_core.h
+++ b/stl/inc/yvals_core.h
@@@ -1760,7 -1760,7 +1761,8 @@@ _EMIT_STL_ERROR(STL1004, "C++98 unexpec
  #define __cpp_lib_constexpr_typeinfo                202106L
  #define __cpp_lib_containers_ranges                 202202L
  #define __cpp_lib_expected                          202211L
 +#define __cpp_lib_flat_map                          202207L
+ #define __cpp_lib_flat_set                          202207L
  #define __cpp_lib_format_ranges                     202207L
  #define __cpp_lib_formatters                        202302L
  #define __cpp_lib_forward_like                      202207L
diff --cc stl/modules/std.ixx
index 5aaad96b,35efad08..3acb3748
--- a/stl/modules/std.ixx
+++ b/stl/modules/std.ixx
@@@ -65,7 -65,7 +65,8 @@@ export module std
  #endif // _HAS_CXX23
  #include <filesystem>
  #if _HAS_CXX23
 +#include <flat_map>
+ #include <flat_set>
  #endif // _HAS_CXX23
  #include <format>
  #include <forward_list>
diff --cc tests/std/include/test_header_units_and_modules.hpp
index a794645a,c9031cfa..7ad2af81
--- a/tests/std/include/test_header_units_and_modules.hpp
+++ b/tests/std/include/test_header_units_and_modules.hpp
@@@ -241,29 -241,12 +241,36 @@@ void test_filesystem() 
  }
  
  #if TEST_STANDARD >= 23
 +void test_flat_map() {
 +    using namespace std;
 +    puts("Testing <flat_map>.");
 +
 +    [[maybe_unused]] constexpr sorted_unique_t unique_tag         = sorted_unique;
 +    [[maybe_unused]] constexpr sorted_equivalent_t equivalent_tag = sorted_equivalent;
 +
 +    constexpr auto simple_truth = [](const auto&) { return true; };
 +
 +    flat_map<int, int> fm;
 +    fm.emplace(42, 172);
 +    fm.emplace(42, 729);
 +    assert(fm.size() == 1);
 +    assert(erase_if(fm, simple_truth) == 1);
 +    assert(fm.empty());
 +
 +    flat_multimap<int, int> fmm;
 +    fmm.emplace(42, 172);
 +    fmm.emplace(42, 729);
 +    assert(fmm.size() == 2);
 +    assert(erase_if(fmm, simple_truth) == 2);
 +    assert(fmm.empty());
 +}
++
+ void test_flat_set() {
+     using namespace std;
+     puts("Testing <flat_set>.");
+ 
+     // FIXME! ADD TEST COVERAGE HERE!
+ }
  #endif // TEST_STANDARD >= 23
  
  void test_format() {
@@@ -1204,7 -1187,7 +1211,8 @@@ void all_cpp_header_tests() 
  #endif // TEST_STANDARD >= 23
      test_filesystem();
  #if TEST_STANDARD >= 23
 +    test_flat_map();
+     test_flat_set();
  #endif // TEST_STANDARD >= 23
      test_format();
      test_forward_list();
diff --cc tests/std/tests/P1502R1_standard_library_header_units/importable_cxx_library_headers.jsonc
index 9d86d337,e9c546b4..d62aa8ec
--- a/tests/std/tests/P1502R1_standard_library_header_units/importable_cxx_library_headers.jsonc
+++ b/tests/std/tests/P1502R1_standard_library_header_units/importable_cxx_library_headers.jsonc
@@@ -23,7 -23,7 +23,8 @@@
      "execution",
      "expected",
      "filesystem",
 +    "flat_map",
+     "flat_set",
      "format",
      "forward_list",
      "fstream",
diff --cc tests/std/tests/P1502R1_standard_library_header_units/test.cpp
index 095c350d,1aa60706..57d47dc0
--- a/tests/std/tests/P1502R1_standard_library_header_units/test.cpp
+++ b/tests/std/tests/P1502R1_standard_library_header_units/test.cpp
@@@ -32,7 -32,7 +32,8 @@@ import <expected>
  #endif // TEST_STANDARD >= 23
  import <filesystem>;
  #if TEST_STANDARD >= 23
 +import <flat_map>;
+ import <flat_set>;
  #endif // TEST_STANDARD >= 23
  import <format>;
  import <forward_list>;
diff --cc tests/std/tests/VSO_0157762_feature_test_macros/test.compile.pass.cpp
index 3e83a6aa,02444ca4..24edabcd
--- a/tests/std/tests/VSO_0157762_feature_test_macros/test.compile.pass.cpp
+++ b/tests/std/tests/VSO_0157762_feature_test_macros/test.compile.pass.cpp
@@@ -353,12 -353,12 +353,18 @@@ STATIC_ASSERT(__cpp_lib_filesystem == 2
  #error __cpp_lib_filesystem is defined
  #endif
  
 +#if _HAS_CXX23
 +STATIC_ASSERT(__cpp_lib_flat_map == 202207L);
 +#elif defined(__cpp_lib_flat_map)
 +#error __cpp_lib_flat_map is defined
 +#endif
 +
+ #if _HAS_CXX23
+ STATIC_ASSERT(__cpp_lib_flat_set == 202207L);
+ #elif defined(__cpp_lib_flat_set)
+ #error __cpp_lib_flat_set is defined
+ #endif
+ 
  #if _HAS_CXX20
  STATIC_ASSERT(__cpp_lib_format == 202304L);
  #elif defined(__cpp_lib_format)
diff --cc tests/std/tests/include_each_header_alone_matrix.lst
index 92769584,4fc17255..329696d4
--- a/tests/std/tests/include_each_header_alone_matrix.lst
+++ b/tests/std/tests/include_each_header_alone_matrix.lst
@@@ -26,7 -26,7 +26,8 @@@ PM_CL="/DMEOW_HEADER=exception
  PM_CL="/DMEOW_HEADER=execution"
  PM_CL="/DMEOW_HEADER=expected"
  PM_CL="/DMEOW_HEADER=filesystem"
 +PM_CL="/DMEOW_HEADER=flat_map"
+ PM_CL="/DMEOW_HEADER=flat_set"
  PM_CL="/DMEOW_HEADER=format"
  PM_CL="/DMEOW_HEADER=forward_list"
  PM_CL="/DMEOW_HEADER=fstream"
```
</details>

### CODE REVIEW

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Looks good! I recommend following this up with a merge from `main` since the addition of `<generator>` will touch the same set of files.

File: stl/inc/xmemory [hunk 1]
```diff
@@ -2619,6 +2619,18 @@ _NODISCARD _Elem* _UIntegral_to_buff(_Elem* _RNext, _UTy _UVal) { // used by bot
...
+_EXPORT_STD struct sorted_unique_t {
+    explicit sorted_unique_t() = default;
+};
+_EXPORT_STD inline constexpr sorted_unique_t sorted_unique{};
+
+_EXPORT_STD struct sorted_equivalent_t {
+    explicit sorted_equivalent_t() = default;
+};
+_EXPORT_STD inline constexpr sorted_equivalent_t sorted_equivalent{};
+#endif // ^^^ _HAS_CXX23 ^^^
```
Comment (frederick-vs-ja) [inline stl/inc/xmemory:2633]: Given now both `<flat_map>` and `<flat_set>` include `<vector>` (for the default adapted containers), is it better to move them to `<vector>` for throughput?

Comment (StephanTLavavej) [reply stl/inc/xmemory:2633]: That's strictly less polluting, so I'm in favor of that unless and until we introduce a dedicated shared header.

**Comment by CaseyCarter** [comment]
> Looks good! I recommend following this up with a merge from `main` since the addition of `<generator>` will touch the same set of files.

Yes, that's the plan.

File: stl/inc/xmemory (hunk 1 above)
Comment (CaseyCarter) [reply stl/inc/xmemory:2633]: Great idea!

**Pull request merged**